### PR TITLE
Support release fd from tcp endpoint

### PR DESCRIPTION
**Pull request by yang-g**
@ctiller

### CODE REVIEW

File: src/core/iomgr/fd_posix.h [hunk 1]
```diff
@@ -111,7 +112,7 @@ grpc_fd *grpc_fd_create(int fd, const char *name);
    notify_on_write.
    MUST NOT be called with a pollset lock taken */
 void grpc_fd_orphan(grpc_exec_ctx *exec_ctx, grpc_fd *fd, grpc_closure *on_done,
```
Comment (ctiller) [inline src/core/iomgr/fd_posix.h]: How about:

```c
void grpc_fd_orphan(..., int *release_fd, ...);
```

With the semantics:
if release_fd == NULL, closes the underlying file descriptor
if release_fd != NULL, does not close the file descriptor, but set *release_fd = underlying_fd before calling on_done

File: src/core/iomgr/tcp_posix.c [hunk 2]
```diff
@@ -468,4 +473,17 @@ grpc_endpoint *grpc_tcp_create(grpc_fd *em_fd, size_t slice_size,
   return &tcp->base;
 }
 
+int grpc_tcp_get_fd(grpc_endpoint *ep) {
```
Comment (ctiller) [inline src/core/iomgr/tcp_posix.c]: I'd prefer that this function didn't exist

File: src/core/iomgr/tcp_posix.c [hunk 3]
```diff
@@ -468,4 +473,17 @@ grpc_endpoint *grpc_tcp_create(grpc_fd *em_fd, size_t slice_size,
...
+  grpc_tcp *tcp = (grpc_tcp *)ep;
+  return tcp->fd;
+}
+
+void grpc_tcp_destroy_and_release_fd(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep,
```
Comment (ctiller) [inline src/core/iomgr/tcp_posix.c]: Like grpc_fd_orphan, how about having an int* argument to this that fetches out the file descriptor before done is called.

File: src/core/iomgr/tcp_posix.c [hunk 4]
```diff
@@ -468,4 +473,17 @@ grpc_endpoint *grpc_tcp_create(grpc_fd *em_fd, size_t slice_size,
...
+                                     grpc_closure *done) {
+  grpc_tcp *tcp = (grpc_tcp *)ep;
```
Comment (ctiller) [inline src/core/iomgr/tcp_posix.c]: Assert that ep->vtable == tcp vtable

File: src/core/iomgr/tcp_posix.c (hunk 2 above)
Comment (yang-g) [inline src/core/iomgr/tcp_posix.c]: Removed.

File: src/core/iomgr/tcp_posix.c (hunk 3 above)
Comment (yang-g) [inline src/core/iomgr/tcp_posix.c]: Done.

**Pull request merged**